android/ui/src/main/java/com/trafi/ui/Badge.kt

```diff
@@ -0,0 +1,354 @@
+    icon: ImageVector? = null,
+    subbadgeIcon: ImageVector? = null,
+    isEnabled: Boolean = true,
+) {
+    ConstraintLayout {
+        Surface(
+            color = if (isEnabled) badge.backgroundColor else constants.disabledColor,
+            contentColor = badge.contentColor ?: constants.defaultContentColor,
+            shape = RoundedCornerShape(badgeType.badgeRounding),
+            modifier = modifier.sizeIn(minHeight = badgeType.badgeHeight)
+                .constrainAs(createRef()) {
+                    bottom.linkTo(parent.bottom)
+                }
+        ) {
+            BadgeFiller(
+                badgeType = badgeType,
+                vector = icon,
+                badge = badge,
+                isStackedBadge = false
+            )
+        }
+        subbadgeIcon?.let {
+            Image(
+                subbadgeIcon,
+                modifier = Modifier
+                    .width(constants.subBadgeIconWidth)
+                    .height(constants.subBadgeIconHeight)
+                    .constrainAs(createRef()) {
+                        bottom.linkTo(parent.bottom)
+                        top.linkTo(parent.bottom)
+                        start.linkTo(parent.end)
+                        end.linkTo(parent.end)
+                    }
+            )
+        }
+    }
+}
+
+
+@OptIn(ExperimentalLayout::class)
+@Composable
+private fun StackedBadge(
+    badge: BadgeInfo,
+    badgeType: BadgeType,
+    modifier: Modifier = Modifier,
+    icon: ImageVector? = null,
+    subbadgeIcon: ImageVector? = null,
+    alternativeBadges: List<BadgeInfo> = listOf(),
+) {
+    ConstraintLayout {
+        val altBadgesSublist = alternativeBadges.take(2)
+        altBadgesSublist.forEachIndexed { index, smallScheduleBadge ->
+            Surface(
+                color = smallScheduleBadge.backgroundColor,
+                contentColor = smallScheduleBadge.contentColor ?: constants.defaultContentColor,
+                border = BorderStroke(constants.borderWidth, color = constants.defaultContentColor),
+                shape = RoundedCornerShape(badgeType.badgeRounding),
+                modifier = modifier.height(badgeType.badgeHeight + (index * 4).dp)
+                    .padding(top = (index * 4).dp)
+            ) {
+                BadgeFiller(
+                    badgeType = badgeType,
+                    vector = icon,
+                    // This is not a mistype, it is a lifehack to make badges same size
+                    badge = badge,
+                    isHiddenLayoutFiller = true,
+                    isStackedBadge = true
+                )
+            }
+        }
+        Surface(
+            color = badge.backgroundColor,
+            contentColor = badge.contentColor ?: constants.defaultContentColor,
+            border = BorderStroke(constants.borderWidth, color = constants.defaultContentColor),
+            shape = RoundedCornerShape(badgeType.badgeRounding),
+            modifier = modifier
```

Comment:
Not clear if this is correct, typically the incoming modifier should be applied to the root / parent layout (i.e. `ConstraintLayout` in this case), but perhaps it works out in this case.